experimental/unicodeset_parser/src/parse.rs

```diff
@@ -0,0 +1,1376 @@
+                for _ in 0..3 {
+                    // is_char_boundary returns true at the latest once exclusive_end == source.len()
+                    if source.is_char_boundary(exclusive_end) {
+                        break;
+                    }
+                    exclusive_end += 1;
+                }
+
+                // exclusive_end is at most source.len() due to str::is_char_boundary and at least 0 by type
+                #[allow(clippy::indexing_slicing)]
+                s.push_str(&source[..exclusive_end]);
+                s.push_str("← ");
+            }
+        }
+        s.push_str("error: ");
+        match kind {
+            ParseErrorKind::UnexpectedChar(c) => {
+                s.push_str(&format!("unexpected character '{}'", c.escape_debug()));
+            }
+            ParseErrorKind::UnknownProperty => {
+                s.push_str("unknown property");
+            }
+            ParseErrorKind::Eof => {
+                s.push_str("unexpected end of input");
+            }
+            ParseErrorKind::Internal => {
+                s.push_str("internal error");
+            }
+            ParseErrorKind::Unimplemented => {
+                s.push_str("unimplemented");
+            }
+            ParseErrorKind::InvalidEscape => {
+                s.push_str("invalid escape sequence");
+            }
+        }
+
+        s
+    }
+
+    /// Returns the [`ParseErrorKind`] of this error.
+    pub fn kind(&self) -> ParseErrorKind {
+        self.kind
+    }
+
+    fn or_with_offset(self, offset: usize) -> Self {
+        match self.offset {
+            Some(_) => self,
+            None => ParseError {
+                offset: Some(offset),
+                ..self
+            },
+        }
+    }
+}
+
+// necessary helper because char::is_whitespace is not equivalent to [:Pattern_White_Space:]
+#[inline]
+fn is_char_pattern_white_space(c: char) -> bool {
+    matches!(
+        c,
+        ('\u{0009}'..='\u{000D}')
+            | '\u{0020}'
+            | '\u{0085}'
+            | '\u{200E}'
+            | '\u{200F}'
+            | '\u{2028}'
+            | '\u{2029}'
+    )
+}
+
+// this ignores the ambiguity between \escapes and \p{} perl properties. it assumes it is in a context where \p is just 'p'
+// returns whether the provided char signifies the start of a literal char (raw or escaped - so \ is a legal char start)
+fn legal_char_start(c: char) -> bool {
+    !(c == '&'
+        || c == '-'
+        || c == '['
+        || c == ']'
+        || c == '{'
+        || c == '}'
+        || is_char_pattern_white_space(c))
+}
+
+#[derive(Debug, Clone, Copy)]
+enum Operation {
+    Union,
+    Difference,
+    Intersection,
+}
+
+// TODO: if UnicodeSetBuilder is staying private, this should have a more generic name like ParseOptions or UnicodeSetParseOptions
+/// Options for parsing a UnicodeSet.
+#[derive(Debug, Clone, Copy, Default)]
+#[non_exhaustive]
+pub struct UnicodeSetBuilderOptions {
```

Comment:
This is good to land in experimental but we still need to land the API updates agreed upon in https://github.com/unicode-org/icu4x/issues/3550